pkg/volume/glusterfs/glusterfs.go

```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
 	}
 	return fmt.Errorf("glusterfs: mount failed: %v", errs)
 }
+
+func (plugin *glusterfsPlugin) NewProvisioner(options volume.VolumeOptions) (volume.Provisioner, error) {
+	if len(options.AccessModes) == 0 {
+		options.AccessModes = plugin.GetAccessModes()
+	}
+	return plugin.newProvisionerInternal(options)
+}
+
+func (plugin *glusterfsPlugin) newProvisionerInternal(options volume.VolumeOptions) (volume.Provisioner, error) {
+	return &glusterfsVolumeProvisioner{
+		glusterfsMounter: &glusterfsMounter{
+			glusterfs: &glusterfs{
+				plugin: plugin,
+			},
+		},
+		options: options,
+	}, nil
+}
+
+type glusterfsClusterConf struct {
+	Glusterep          string `json:"endpoint"`
+	GlusterRestvolpath string `json:"path"`
+	GlusterRestUrl     string `json:"resturl"`
+	GlusterRestAuth    bool   `json:"restauthenabled"`
+	GlusterRestUser    string `json:"restuser"`
+	GlusterRestUserKey string `json:"restuserkey"`
+}
+
+type glusterfsVolumeProvisioner struct {
+	*glusterfsMounter
+	*glusterfsClusterConf
+	options volume.VolumeOptions
+}
+
+var clusterconf = new(glusterfsClusterConf)
+
+func (plugin *glusterfsPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	return plugin.newDeleterInternal(spec)
+}
+
+func (plugin *glusterfsPlugin) newDeleterInternal(spec *volume.Spec) (volume.Deleter, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.Glusterfs == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.Spec.Glusterfs is nil")
+	}
+	return &glusterfsVolumeDeleter{
+		glusterfsMounter: &glusterfsMounter{
+			glusterfs: &glusterfs{
+
+				volName: spec.Name(),
+				plugin:  plugin,
+			},
+
+			path: spec.PersistentVolume.Spec.Glusterfs.Path,
+		}}, nil
+}
+
+type glusterfsVolumeDeleter struct {
+	*glusterfsMounter
+	*glusterfsClusterConf
+}
+
+func (d *glusterfsVolumeDeleter) GetPath() string {
+	name := glusterfsPluginName
+	return d.plugin.host.GetPodVolumeDir(d.glusterfsMounter.glusterfs.pod.UID, strings.EscapeQualifiedNameForDisk(name), d.glusterfsMounter.glusterfs.volName)
+}
+
+func (d *glusterfsVolumeDeleter) Delete() error {
+	return d.DeleteVolume()
+}
+
+func (d *glusterfsVolumeDeleter) DeleteVolume() error {
```

Comment:
What's the difference between `glusterfsVolumeDeleter.Delete` and `glusterfsVolumeDeleter.DeleteVolume`? Should it be the same method?